### DISCUSSION TITLE
How the dynamic library system works

### DISCUSSION THREAD

**Discussion by MyelinsheathXD** [ezEngine/ezEngine#550] [General]
Hello folks! I want to create a game engine like you guys did.

Currently I am struggling with creating an dynamic  architecture of the engine that controls external dynamic or shared library. Where can I learn this tech ?. I know that Ez can add several plugins into the engine. how those systems works? 
Specifically 
1. building a dll using cmake that can be inserted into the engine , and then use it as components in the editor
2. handling all components inside the engine that is the core engine

**Comment by jankrassnigg** [comment]
Hi, please don't cross post the same question both here AND in the Discord server.

For reference, here is what others already answered, please continue the discussion only here.

For windows you'll use these family of functions:
https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibrarya
On unix-like: https://tldp.org/HOWTO/Program-Library-HOWTO/dl-libraries.html
For building a dll or so binary, just create a cmake shared library
like add_library(LibName SHARED sources)
You'll also want to export those functions that you want to access from other libraries
There's more involved but once you start reading docs on it you'll come across the specifics.

**Comment by jankrassnigg** [comment]
If you want to learn how EZ does it's plugin loading, look at [ezPlugin](https://ezengine.github.io/api-docs/dd/d46/classez_plugin.htm). There is quite some documentation in the code, explaining some of the issues. Some more info is also here: http://ezengine.net/pages/docs/custom-code/cpp/engine-plugins.html

For finding types across DLLs, we rely heavily on ezEnumerable, which uses patterns that work with DLLs. For example ezRTTI, our reflection infrastructure, is also build upon that.

A lot of my knowledge about what works with DLLs and what not, comes from the book [Imperfect C++](https://www.amazon.com/Imperfect-Practical-Solutions-Real-Life-Programming/dp/0321228774/ref=sr_1_1?crid=CVVKHAJXPDMV&keywords=imperfect+c%2B%2B&qid=1641713752&sprefix=imperfect+c%2B%2B%2Caps%2C149&sr=8-1)

That's all more advanced though. First try the basics (building a DLL and an EXE, load the DLL dynamically into the EXE). There are a lot of resources online that show how to do that.
